---
title: "Figs"
author: "JBone"
date: "2025-02-03"
output: html_document
---


```{r setup, include=FALSE}
setwd("/Users/jordanbone/Documents/GitHub/IAV-Mach-Learning")
source("MegaLibrary.R")

reference_table <- read.csv("RefTable.csv")
rep_uids <- read.csv("USED UIDS.txt")
uid_ref <- read.csv("USED UIDS.txt")
uid_ref <- inner_join(uid_ref,reference_table[,c(1,14,15)]) %>% distinct()
```
# Data Collection
```{r, eval=FALSE}
a <- subset(reference_table,Protein %in% PRT & UID %in% uid_ref$UID) %>% select(UID,ntAcc,Sample,Classification,Date)
names(a) <- c("UID","Acc","Sample","Classification","Date")
b <- subset(reference_table,Protein %in% PRT & UID %in% uid_ref$UID) %>% select(UID,ptAcc.x,Sample,Classification,Date)
names(b) <- c("UID","Acc","Sample","Classification","Date")
rbind(a,b) %>% write.csv("feats/combo/Timings.csv",row.names = F)
```

```{r Reference Builder, eval=FALSE}
clu <- list.files("feats/combo",pattern = ".csv",full.names = T,recursive = T)
cls <- data.frame()
for(i in 1:length(clu)){
  clts <- data.frame(ptAcc=read.csv(clu[i])[,1] %>% str_split_i(":",1),
                     prt=clu[i] %>% str_split_i("\\/",2) %>% str_split_i("_",1))
  cls <- rbind(cls,clts)
}

cls2 <- left_join(cls,reference_table, relationship = "many-to-many")
```

```{r Reference Builder, eval=FALSE}
####
usrs <- read_xlsx("/Users/jordanbone/Documents/IAV.AI/Used Seqs.xlsx")[,c(2:10)]
us2 <- usrs %>% pivot_longer(2:9, names_to = "Protein", values_to = "ptAcc2")
reference_table <- left_join(reference_table,us2,by=c("UID","Protein"))

for(i in 1:length(reference_table$UID)){
#   if(reference_table[i,12]!="08NEP" & reference_table[i,12]!="07M2" & reference_table[i,12]!="02PB1-F2" & reference_table[i,12]!="03PA-X"){
#     if(reference_table[i,1] %in% us2$UID){
#       subs <- subset(us2,UID %in% reference_table[i,1] & Protein == reference_table[i,12])
#       reference_table[i,11] <- ifelse(reference_table[i,11]!=subs$ptAcc,subs$ptAcc,reference_table[i,11])
      #     # reference_table[i,9] <- ifelse(reference_table[i,9]!=subs$ntAcc|is.na(reference_table[i,9]),subs$ntAcc,reference_table[i,9])
      #     # reference_table[i,14] <- ifelse(reference_table[i,14]!=subs$Subtype|is.na(reference_table[i,14]),subs$Subtype,reference_table[i,14])
#     }}
  reference_table[i,11] <- ifelse(reference_table[i,11]!=reference_table[i,16] &!is.na(reference_table[i,16]),reference_table[i,16],reference_table[i,11])
  per.done(i)
}
beep(2)
write.csv(reference_table[,-16],"RefTable.csv",row.names = F)
```

```{r}
mam <- subset(reference_table,Class=="Mammalia") %>%
  group_by(Order,Family,Subtype) %>% summarise(N=length(Subtype)) %>%
  subset(Family %in% c("Hominidae","Suidae","Canidae","Equidae","Phyllostomidae"))

ggplot(subset(mam,N>19),aes(Family,log10(N),fill=Subtype))+geom_col(position="dodge")+
  geom_text(aes(label=Subtype,group=Subtype),position = position_dodge(width = .9),size=3)
# ggsave("Mammal Subtypes.pdf",width=13,height=8)
```

```{r}
fetc <- list.files("feats",pattern = ".csv",full.names = T,recursive = F)
for(i in 1:length(fetc)){
  FeatParam <- str_split_i(fetc[i],"\\/",2) %>% str_replace_all(".csv","")
  SegFeat <- paste0(str_split_i(FeatParam,"_",1),"_",str_split_i(FeatParam,"_",8))
  assign(SegFeat, read.csv(fetc[i]))
}

ha_ctdc$Seg <- "HA"
ha_ctdc$Feat <- "CTDc"
```

```{r}
# subset(uid_ref,!is.na(Subtype)) %>% group_by(Subtype) %>% summarise(Num=length(Subtype)) %>%  ggplot(aes(x=reorder(Subtype,-Num),y=Num,fill=Num<25))+ geom_bar(stat="identity")+ geom_label(aes(x=20,y=200,label="20 of which appear\nmore than 25 times"),fill="#F8766D")+ theme(legend.position = "none",axis.text.x = element_text(angle = 90))+labs(title = "95 Unique Subtypes")

subset(uid_ref,!is.na(Subtype)) %>% group_by(Subtype) %>% summarise(Num=length(Subtype)) %>%  ggplot(aes(x=reorder(Subtype,-Num),y=Num,fill=Subtype %in% held_subtypes))+ geom_bar(stat="identity")+ geom_label(aes(x=20,y=200,label="23 of which appear\nmore than 25 times"),fill="#53CCCD")+
  geom_segment(aes(x=70,xend=59.5,y=310,yend=20),arrow = arrow(angle = 30, length = unit(0.2, "cm"), type = "closed"),colour="#53CCCD")+
  geom_segment(aes(x=70,xend=79.5,y=310,yend=20),arrow = arrow(angle = 30, length = unit(0.2, "cm"), type = "closed"),colour="#53CCCD")+
  geom_label(aes(x=70,y=300,label="Plus two bat\nsequences"),fill="#53CCCD")+
  theme(legend.position = "none",
        text=element_text(size=20), #change font size of all text
        axis.title=element_text(size=20), #change font size of axis titles
        plot.title=element_text(size=20), #change font size of plot title
        legend.text=element_text(size=20), #change font size of legend text
        legend.title=element_text(size=20),
        axis.text.x = element_text(size=8,angle = 90),)+
  labs(title = "95 Unique IAV Subtypes",x="Subtypes",y="Quantity")
ggsave("Used Subtypes.pdf",width=33,height=19,units = "cm")
```

```{r}
# ssnl <- read_xlsx("Human Seasonals.xlsx",sheet="Sheet5")
# ggplot(ssnl,aes(Release_Date,WGS,colour=Genotype))+geom_point(alpha=0.7)
```

```{r}
mod_cls <- read.csv("Sequences/Clusters.csv")
ggplot(mod_cls,aes(Protein,Clusters,fill=Group))+geom_col(position="dodge")+
  facet_wrap(~Identity,scales = "free")+
  theme(legend.position = "bottom",axis.text.x = element_text(angle = 90))

mod_cls %>% pivot_wider(names_from = c("Group","Identity"),values_from = "Clusters") 
# %>% kbl(col.names = c("Protein","100%","95%","85%","75%","100%","95%","85%","75%"),align='c',caption="Number of representative sequences from each group") %>% kable_styling(full_width = F) %>% add_header_above(c(" "=1,"Avian"=4,"Mammalian"=4))

# a minimum coverage (option -c [0,1], which is defined by the number of aligned residue pairs divided by either the maximum of the length of query/centre and target/non-centre sequences alnRes/max(qLen,tLen) (default mode, --cov-mode 0), or by the length of the target/non-centre sequence alnRes/tLen (--cov-mode 1), or by the length of the query/centre alnRes/qLen (--cov-mode 2).

# a minimum sequence identity (--min-seq-id [0,1]) with option --alignment-mode 3 defined as the number of identical aligned residues divided by the number of aligned columns including internal gap columns, or, by default, defined by a highly correlated measure, the equivalent similarity score of the local alignment (including gap penalties) divided by the maximum of the lengths of the two locally aligned sequence segments. The score per residue equivalent to a certain sequence identity is obtained by a linear regression using thousands of local alignments as training set.
```

```{r Sequence Diversity, include=FALSE}
# haBifas <- seqinr::read.fasta("Sequences/Avian/ha_cluster_95_c80_cov0_rep_seq.fasta")
# haBi <- data.frame("Base"=1:567,"Entropy"=Bios2cor::entropy(haBifas),row.names = NULL)
# haMafas <- seqinr::read.fasta("Sequences/Mammals/04HA/HA_cluster_95.fasta")
# haMa <- data.frame("Base"=1:566,"Entropy"=Bios2cor::entropy(haMafas),row.names = NULL)
# 
# ggplot(haBi,aes(Base,Entropy))+geom_line(colour="indianred",alpha=0.8)+geom_line(data=haMa,colour="slateblue",alpha=0.8)+
#   geom_label(aes(300,0.58,label="Avian\nH = 93.13"),colour="indianred")+
#   geom_label(aes(400,0.4,label="Mammalian\nH = 90.30"),colour="slateblue")+labs(title="Sitewise Diversity in Avian and Mammalian Haemagglutinin Clusters")
```

```{r Lolliplots, include=FALSE}
# HASNPcon <- c(144,467)
# sample.HAcon <- GRanges("HA", IRanges(HASNPcon, width=1, names=c("Gly 144 Asp","Arg 467 Ser")))
# features.HA <- GRanges("HA", IRanges(c(1,1,17,52,276,330,347,514),width=c(565,16,51,223,53,16,166,20),names=c("HA","Signal Peptide","HA1 - Stalk","HA1 - Head","HA1 - Stalk","Fusion Peptide","HA2 - Stalk","Transmembrane Domain")))
# features.HA$fill <- c("ivory","tan","plum3","purple3","plum3","firebrick2","lightslateblue","sienna1")
# features.HA$height <- 0.08
# sample.HAcon$color <- rep("dodgerblue2",2)
# sample.HAcon$shape <- rep("circle",2)
# trackViewer::lolliplot(sample.HAcon,features.HA,legend = legend)
# grid.text("HA Segment", x=.5, y=.98, just="top",gp=gpar(cex=1.5, fontface="bold"))
```

```{r}
# num_clus <- read.csv("Sequences/Clustering Tests.csv")
# subset(num_clus,Identity!=100 & Group=="Avian")[,-1] %>% pivot_wider(names_from = c("Identity","Coverage"), values_from = "Count") %>% cbind(Total=c(15507,15507,15447,15447,15441,15441,15380,15380,14682,14682,15442,15442,15335,15335,15124,15124,15228,15228,15293,15293)) %>%  kbl(col.names = c("Protein","Mode",rep(c("c 50","c 70","c 80"),5),"Total"), caption="Proteins of Avian Viruses") %>% kable_styling(full_width = F) %>% add_header_above(c(" "=1," "=1,"75% Identity"=3,"85% Identity"=3,"90% Identity"=3,"95% Identity"=3,"99% Identity"=3," "=1)) %>% column_spec(12:14,background = "palegreen") %>% collapse_rows(c(1:18))

# subset(num_clus,Identity!=100 & Group=="Mammal")[,-1] %>% pivot_wider(names_from = c("Identity","Coverage"), values_from = "Count") %>% cbind(Total=c(21880,21880,21829,21829,21856,21856,22106,22106,21865,21865,21845,21845,21791,21791,21770,21770,21760,21760,21737,21737)) %>% arrange(Protein) %>%  kbl(col.names = c("Protein","Mode",rep(c("c 50","c 70","c 80"),5),"Total"), caption="Proteins of Mammalian Viruses") %>% kable_styling(full_width = F) %>% add_header_above(c(" "=1," "=1,"75% Identity"=3,"85% Identity"=3,"90% Identity"=3,"95% Identity"=3,"99% Identity"=3," "=1)) %>% column_spec(12:14,background = "palegreen") %>% collapse_rows(c(1:18))

# subset(cls2,Protein!="02PB1-F2"&Protein!="03PA-X"&Protein!="07M2"&Protein!="08NEP") %>% group_by(Protein,Subtype,Class) %>% summarise(N=length(UID)) %>% pivot_wider(names_from = Protein,values_from = N) %>% 
  # write.csv("Cluster Selection.csv",row.names = F)
  # kbl() %>% kable_styling(full_width = F)

cls3 <- subset(cls2,Protein!="02PB1-F2"&Protein!="03PA-X"&Protein!="07M2"&Protein!="08NEP") %>% group_by(Subtype,Class,Family,Superorder,Infraclass) %>% summarise(N=length(UID))

ggplot(subset(cls3,Class=="Mammalia"),aes(Family,N,fill=Subtype))+
  geom_col(position="dodge")+geom_text(aes(y=N+20,label=Subtype),check_overlap = T,position = position_dodge(width = .9),size=2.5)+
  theme(legend.position = "none")

ggplot(subset(cls3,Class=="Aves"),aes(Family,N,fill=Subtype))+geom_col(position="dodge")+geom_text(aes(y=N*.5,label=Subtype),check_overlap = T,position = position_dodge(width = .9),size=2.5)+
  theme(legend.position = "none")+facet_wrap(Infraclass~Superorder, scales = "free")

ggplot(subset(cls3,Superorder=="Galloanserae"),aes(Family,N,fill=Subtype))+geom_col(position="dodge")+geom_text(aes(y=N*.5,label=Subtype),check_overlap = T,position = position_dodge(width = .9),size=2.5)+
  theme(legend.position = "none")
ggplot(subset(cls3,Superorder=="Neoaves"),aes(Family,N,fill=Subtype))+geom_col(position="dodge")+geom_text(aes(y=N*.5,label=Subtype),check_overlap = T,position = position_dodge(width = .9),size=2.5)+
  theme(legend.position = "none",axis.text.x = element_text(angle = 90))

cls3 %>% group_by(Subtype) %>% summarise(sn=sum(N)) %>% subset(sn>25)

ggplot(cls3,aes(x=reorder(Subtype,-N),y=N,fill=ifelse(Class=="Aves",Class,Family)))+geom_col(position="dodge")+theme(legend.position = "bottom",axis.text.x = element_text(angle = 90))+labs(x="Subtype",fill="Group")+geom_vline(xintercept = 58.5, alpha=0.4, linetype="longdash")
ggsave("Subtype Frequencies.pdf",width=33,height=19,units = "cm")

ggplot(subset(cls3,N>=50),aes(x=reorder(Subtype,-N),y=N,fill=ifelse(Class=="Aves",Class,Family)))+geom_col(position="dodge")+theme(legend.position = "bottom",axis.text.x = element_text(angle = 90))+labs(x="Subtype",fill="Group")+geom_vline(xintercept = 58.5, alpha=0.4, linetype="longdash")
ggsave("Subtype Frequencies zoom.pdf",width=33,height=19,units = "cm")

cls3$Classification <- ifelse(cls3$Class=="Mammalia",cls3$Family,str_replace_all(cls3$Superorder,"Aves","Galloanserae"))
cls4 <- cls3 %>% group_by(Subtype,Classification) %>% summarise(N=sum(N))
ggplot(cls4,aes(x=reorder(Subtype,-N),y=N,fill=Classification))+geom_col()+theme(legend.position = "bottom",axis.text.x = element_text(angle = 90))
```

# Sequence Cleaning
```{r Cluster Tests, include=FALSE}
stack_res <- read_xlsx("Comp/HA test results.xlsx",sheet="Stacks")
ensemb_res <- read_xlsx("Comp/HA test results.xlsx",sheet="Ensemble")

ggplot(subset(stack_res,Model!="ensemble"),aes(paste(Coverage,Mode),Importance,colour=Model))+
  geom_point()
ggplot(stack_res,aes(paste(Coverage,Mode),ymin=ROC-sdROC,y=ROC,ymax=ROC+sdROC,fill=Model))+
  geom_crossbar(position="dodge")

# stack_res %>% kbl() %>% kable_styling(full_width = F) %>% collapse_rows(4:5)
# ensemb_res %>% kbl() %>% kable_styling(full_width = F) %>% collapse_rows(4:5)
```

```{r Model Check}
mo_list <- list.files("Models/",pattern=".rds",full.names = T)
mod_stats_2 <- data.frame()
mod_stats_M <- data.frame()
for(i in 1:length(mo_list)){
  tmp_mo <- readRDS(mo_list[i])
  if(str_split_i(mo_list[i],"_",5)=="Mclass.rds"){
    temp_df_M <- data.frame(Protein=str_split_i(mo_list[i],"_",1) %>% str_split_i("\\/",3),
                            Feature=str_split_i(mo_list[i],"_",2),
                            Holdout=str_split_i(mo_list[i],"_",3),
                            ModelType=str_split_i(mo_list[i],"_",5) %>% str_split_i("l",1),
                            Samples=tmp_mo$finalModel$num.samples,
                          subset(tmp_mo$results,min.node.size==tmp_mo$finalModel$tuneValue$min.node.size))
    mod_stats_M <- rbind(mod_stats_M,temp_df_M)
  }
  else{
    temp_df_2 <- data.frame(Protein=str_split_i(mo_list[i],"_",1) %>% str_split_i("\\/",3),
                            Feature=str_split_i(mo_list[i],"_",2),
                            Holdout=str_split_i(mo_list[i],"_",3),
                            ModelType=str_split_i(mo_list[i],"_",5) %>% str_split_i("l",1),
                            Samples=tmp_mo$finalModel$num.samples,
                            subset(tmp_mo$results,min.node.size==tmp_mo$finalModel$tuneValue$min.node.size))
    mod_stats_2 <- rbind(mod_stats_2,temp_df_2)
  }
  ifelse(is.wholenumber(i/100),print(i),"")
}

ggplot(mod_stats_M,aes(Feature,Accuracy,colour=Holdout))+geom_point(alpha=0.6)+facet_wrap(~Protein)
ggplot(mod_stats_M,aes(Mean_Specificity,Mean_Sensitivity,colour=Holdout))+geom_point(alpha=0.6)+facet_wrap(~Feature)

ggplot(mod_stats_2,aes(Feature,ROC,colour=Holdout))+geom_point(alpha=0.6)+facet_wrap(~Protein)
ggplot(mod_stats_2,aes(Spec,Sens,colour=Holdout))+geom_point(alpha=0.6)+facet_wrap(~Feature)
```


```{r Model Outputs}
# mos_out <- rbind(data.frame(protein="07M1",feature=str_split_i(names(summary(M1_complete)$imp),"\\.",1),imp=summary(M1_complete)$imp,summary(M1_complete)$results[2:7],row.names = NULL),rbind(data.frame(protein="04HA",feature=str_split_i(names(summary(HA_complete)$imp),"\\.",1),imp=summary(HA_complete)$imp,summary(HA_complete)$results[2:7],row.names = NULL),rbind(data.frame(protein="06NA",feature=str_split_i(names(summary(NA_complete)$imp),"\\.",1),imp=summary(NA_complete)$imp,summary(NA_complete)$results[2:7],row.names = NULL),rbind(data.frame(protein="03PA",feature=str_split_i(names(summary(PA_complete)$imp),"\\.",1),imp=summary(PA_complete)$imp,summary(PA_complete)$results[2:7],row.names = NULL),rbind(data.frame(protein="08NS1",feature=str_split_i(names(summary(NS1_complete)$imp),"\\.",1),imp=summary(NS1_complete)$imp,summary(NS1_complete)$results[2:7],row.names = NULL),rbind(data.frame(protein="02PB1",feature=str_split_i(names(summary(PB1_complete)$imp),"\\.",1),imp=summary(PB1_complete)$imp,summary(PB1_complete)$results[2:7],row.names = NULL),rbind(data.frame(protein="05NP",feature=str_split_i(names(summary(NP_complete)$imp),"\\.",1),imp=summary(NP_complete)$imp,summary(NP_complete)$results[2:7],row.names = NULL),data.frame(protein="01PB2",feature=str_split_i(names(summary(PB2_complete)$imp),"\\.",1),imp=summary(PB2_complete)$imp,summary(PB2_complete)$results[2:7],row.names = NULL))))))))

# data.frame(protein="01PB2",feature=str_split_i(names(summary(PB2_complete)$imp),"\\.",1),imp=summary(PB2_complete)$imp,summary(PB2_complete)$results[2:7],row.names = NULL)

mos_out <- read.csv("Comp/Model Outputs.csv")
ggplot(mos_out,aes(feature,ymin=value-sd,y=value,ymax=value+sd,colour=imp))+
  geom_pointrange()+facet_wrap(~protein,ncol=4)+scale_colour_gradient(low="grey",high="red")+theme(legend.position = "bottom",axis.text.x = element_text(angle = 45))
```

```{r Subtype Distribution, include=FALSE}
sbd <- cls2 %>% group_by(Order,Class,Superorder,Subtype) %>% summarise(N=length(UID))
sbd$Classification <- ifelse(sbd$Class=="Mammalia",sbd$Order,sbd$Superorder)
# sbd %>% group_by(Classification,Subtype) %>% summarise(N=sum(N)) %>% pivot_wider(names_from = "Classification",values_from = N) %>% write.csv("Subtype Distribution.csv",row.names = F)

reference_table[,c(1,14,15,11,12)] %>% subset(Protein!="02PB1-F2"&Protein!="03PA-X") %>% group_by(UID,Subtype,Classification,Protein) %>% summarise(N=length(ptAcc)) %>% pivot_wider(names_from = Protein, values_from = N)

# dupes <- reference_table[,c(1,14,15,11,12)] %>% subset(Protein!="02PB1-F2"&Protein!="03PA-X") |> dplyr::summarise(n = dplyr::n(), .by = c(UID, Subtype,Classification, Protein)) |> dplyr::filter(n > 1L)

ax <- reference_table[,c(1,14,15,11,12)] %>% subset(Protein!="02PB1-F2"&Protein!="03PA-X"&Protein!="08NEP"&Protein!="07M2") 
ax$Rep <- ifelse(ax$ptAcc %in% cls$ptAcc,"Rep","Not")

badax <- ax |> dplyr::summarise(n = dplyr::n(), .by = c(UID, Subtype, Classification, Rep, Protein)) |> dplyr::filter(n > 1L)

for(i in 1:length(ax$UID)){
  ax[i,4] <- ifelse(ax[i,6]=="Rep",cell_spec(ax[i,4], background = "forestgreen"),ax[i,4])
}
# ax$ptAcc <- cell_spec(ax$ptAcc, background = ifelse(ax$Rep=="Rep","forestgreen","white"))

subset(ax,!UID %in% badax$UID)[,-6] %>% group_by(UID,Subtype,Classification) %>% pivot_wider(names_from = Protein, values_from = ptAcc) %>% kbl(escape=F) %>% kable_styling() %>% save_kable(file = "table1.html")
# %>% write.csv("Full data.csv",row.names = F)

mod_stats <- read.csv("Models/Mod Stats.csv")
subset(mod_stats,FeatSet=="ctdd") %>% pivot_longer(cols=4:5,names_to = "Class",values_to = "N") %>% ggplot(aes(Holdout,N,colour=Class))+geom_point(alpha=0.6)
subset(mod_stats,Avian+Mammal!=2325)

mostat <- read.csv("MoStat.csv")
ggplot(subset(mostat,F=="ctdd"),aes(str_replace_all(X,"ex",""),N))+geom_point(alpha=0.6)+facet_wrap(~P,scales="free",nrow=2)+theme(legend.position = "bottom",axis.text.x = element_text(angle = 90))+labs(x="Subtype")
ggsave("Num Holdouts.pdf",width=50,height=19,units = "cm")

ggplot(subset(mostat,F=="ctdd"&P=="04ha"),aes(str_replace_all(X,"ex",""),N,colour=N==max(N)))+geom_point(alpha=0.9)+theme(legend.position = "bottom",axis.text.x = element_text(angle = 90))+labs(x="Subtype")
ggsave("HA Holdouts.pdf",width=30,height=19,units = "cm")
ha2324 <- subset(mostat,P=="ha"&N==2324&F=="2mer")$X

ggplot(subset(mostat,F=="ctdd"&P=="ha"&N<2300),aes(str_replace_all(X,"ex",""),y=max(mostat$N)-N))+geom_point(alpha=0.9)+labs(x="Excluded Subtype",y="Number of Sequences")
```

# Model Stats
```{r}
# all_mods <- list.files("Comp",pattern=".rds",full.names = T)
# mod_stats <- data.frame()
# for(i in 1:length(all_mods)){
# # for(i in 1:2){
#   STY <- all_mods[i] %>% str_split_i("_",2)
#   PRT <- all_mods[i] %>% str_split_i("_",1) %>% str_split_i("\\/",2)
#   MTY <- all_mods[i] %>% str_split_i("_",3) %>% str_sub(1,1)
#   MOD <- summary(readRDS(all_mods[i]))
#   mod_df <- data.frame(STY,PRT,MTY,MOD$imp)
#   mod_df$MDL <- rownames(mod_df) %>% str_split_i("\\.",1)
#   mod_df$HST <- rownames(mod_df) %>% str_replace("_mer.r","mer.r") %>% str_split_i("_",2)
#   mod_stats <- rbind(mod_stats,mod_df)
# }
```

```{r}
mod_imp <- read.csv("Models/Model Importance.csv")

a2 <- ggplot(subset(mod_imp,MTY=="2"),aes(MDL,MOD.imp))+geom_violin(scale="area")+
  theme(axis.text.x = element_text(angle = 90))
am <- ggplot(subset(mod_imp,MTY=="m"),aes(MDL,MOD.imp))+geom_violin(scale="area")+
  facet_wrap(~HST)+theme(axis.text.x = element_text(angle = 90))
ggarrange(a2,am)
ggsave("Model Importance.pdf",width=33,height=19,units = "cm")
```

## Confusion Matrices
```{r}
master_conf <- readRDS("Comp/ConfMatrix Master Multiclass.rds")

master_conf$table %>% as.data.frame() %>% ggplot(aes(Prediction,Reference,fill=Freq/sum(master_conf$table)))+
  geom_tile(colour="grey10")+geom_text(aes(label=ifelse(Freq<=0," ",percent(Freq/sum(master_conf$table),accuracy = 0.001))))+
  scale_fill_gradient(low="white",high="forestgreen",na.value = "grey50")+guides(fill=guide_legend(title="Freq %"))
ggsave("Mislabelling.pdf",width=33,height=19,units = "cm")

classes <- master_conf$byClass %>% as.data.frame() %>% rownames_to_column()
# classes$rowname <- str_split_i(classes$rowname,":",2) %>% str_trim
classes$Host <- c("Avian","Dog","Horse","Human","Bat","Swine")
m1 <- ggplot(classes,aes(Sensitivity,Specificity,colour=Host))+geom_point(size=3)+ylim(0,1)
m2 <- ggplot(classes,aes(Host,Precision,colour=Host))+geom_point(size=3)+ylim(0,1)
m3 <- ggplot(classes,aes(Host,`Balanced Accuracy`,colour=Host))+geom_point(size=3)+ylim(0,1)
m4 <- ggplot(classes,aes(`Pos Pred Value`,`Neg Pred Value`,colour=Host))+geom_point(size=3)+ylim(0,1)
ggarrange(m1,m4,m3,m2,common.legend = T,legend="bottom")
ggsave("Model Diag.pdf",width=33,height=19,units = "cm")

master_conf2 <- readRDS("Comp/ConfMatrix Master 2class.rds")
master_conf2$table %>% as.data.frame() %>% ggplot(aes(Prediction,Reference,fill=Freq/sum(master_conf$table)))+
  geom_tile(colour="grey10")+geom_text(aes(label=ifelse(Freq<=0," ",percent(Freq/sum(master_conf$table),accuracy = 0.001))))+
  scale_fill_gradient(low="white",high="forestgreen",na.value = "grey50")+guides(fill=guide_legend(title="Freq %"))
ggsave("Figures & Presentables/Mislabelling 2c.pdf",width=33,height=19,units = "cm")

overall <- master_conf$overall %>% t() %>% as.data.frame() %>% rbind(master_conf2$overall %>% t() %>% as.data.frame()) %>% cbind() %>% data.frame("Model"=c("Multiclass","Dualclass"))
ggplot(overall,aes(x=Model,ymin=AccuracyLower,y=Accuracy,ymax=AccuracyUpper))+
  geom_pointrange(size=0.2)
ggsave("Figures & Presentables/Overall Conf.pdf",width=33,height=19,units = "cm")
```

### Protein by Protein
```{r,eval=FALSE}
  overall_2c <- c()
  byClass_2c <- c()
  overall_Mc <- c()
  byClass_Mc <- c()
for (PRT in c("PB2","PB1","PA","HA","NP","NA","M1","NS1")){
  cm2c <- readRDS(paste0("Comp/",PRT," 2class ConfMatrix.rds"))
  cmMc <- readRDS(paste0("Comp/",PRT," Mclass ConfMatrix.rds"))
  
  # cm2c$table
  ovr2 <- data.frame(PRT,t(cm2c$overall))
  byc2 <- data.frame(PRT,Val=cm2c$byClass) %>% rownames_to_column()
  
  # cmMc$table
  ovrM <- data.frame(PRT,t(cmMc$overall))
  bycM <- data.frame(PRT,cmMc$byClass) %>% rownames_to_column()
  
  overall_2c <- bind_rows(overall_2c,ovr2)
  byClass_2c <- bind_rows(byClass_2c,byc2)
  overall_Mc <- bind_rows(overall_Mc,ovrM)
  byClass_Mc <- bind_rows(byClass_Mc,bycM)
}

overall_2c
byClass_2c <- byClass_2c %>% pivot_wider(names_from = rowname,values_from = Val)
overall_Mc
byClass_Mc$rowname <- byClass_Mc$rowname %>% str_split_i(":",2) %>% str_trim()
```

```{r}
overall_results <- read.csv("Comp/Model results.csv")
tc_byc <- read.csv("Comp/Two class breakdown.csv")
mc_byc <- read.csv("Comp/Multiclass breakdown.csv")

overall_results$Accuracy <- percent(overall_results$Accuracy,accuracy = 0.01)
overall_results[,c(2,1,3)] %>% arrange(Protein) %>% kbl() %>% kable_styling(full_width = F) %>% collapse_rows(1)
```


```{r}
ggplot(overall_results,aes(Protein,ymin=AccuracyLower,y=Accuracy,ymax=AccuracyUpper))+geom_pointrange()+facet_wrap(~ModelType)
ggplot(overall_results,aes(Protein,Kappa))+geom_point()+facet_wrap(~ModelType)

ggplot(tc_byc,aes(Sensitivity,Specificity,colour=Protein))+geom_point()
ggplot(tc_byc,aes(Pos.Pred.Value,Neg.Pred.Value,colour=Protein))+geom_point()
ggplot(tc_byc,aes(Protein,Precision))+geom_point()
ggplot(tc_byc,aes(Protein,F1))+geom_point()
ggplot(tc_byc,aes(Protein,Balanced.Accuracy))+geom_point()

ggplot(mc_byc,aes(Sensitivity,Specificity,colour=Protein))+geom_point(alpha=0.6)+facet_wrap(~Host)
ggplot(mc_byc,aes(Pos.Pred.Value,Neg.Pred.Value,colour=Protein))+geom_point()+facet_wrap(~Host)
ggplot(mc_byc,aes(Protein,Precision,colour=Host))+geom_jitter(height=0,width=0.2)
ggplot(mc_byc,aes(Protein,F1,colour=Host))+geom_point()
ggplot(mc_byc,aes(Protein,Balanced.Accuracy,colour=Host))+geom_point()
```

```{r}
ggplot(mc_byc,aes(Protein,Host,fill=F1))+geom_tile()+geom_text(aes(label=percent(F1,accuracy=0.01)))+
  scale_fill_gradient(low="grey90",high="forestgreen",na.value = "white")+ theme(legend.position="none")+ labs(title = "F1 Values of Multiclass Models")

ggplot(mc_byc,aes(Protein,Host,fill=Balanced.Accuracy))+geom_tile()+scale_fill_gradient(low="grey90",high="forestgreen")+geom_text(aes(label=percent(F1,accuracy=0.01)))+
  scale_fill_gradient(low="grey90",high="forestgreen",na.value = "white")+ theme(legend.position="none")+ labs(title = "Balanced Accuracy of Multiclass Models")
```


# VarImp
```{r}
PB2v <- read.csv("Varimp/varimp_PB2_mclass.csv") %>% mutate(Protein="01PB2")
PB1v <- read.csv("Varimp/varimp_PB1_mclass.csv") %>% mutate(Protein="02PB1")
PAv <- read.csv("Varimp/varimp_PA_mclass.csv") %>% mutate(Protein="03PA")
HAv <- read.csv("Varimp/varimp_HA_mclass.csv") %>% mutate(Protein="04HA")
NPv <- read.csv("Varimp/varimp_NP_mclass.csv") %>% mutate(Protein="05NP")
NAv <- read.csv("Varimp/varimp_NA_mclass.csv") %>% mutate(Protein="06NA")
M1v <- read.csv("Varimp/varimp_M1_mclass.csv") %>% mutate(Protein="07M1")
NS1v <- read.csv("Varimp/varimp_NS1_mclass.csv") %>% mutate(Protein="07NS1")

var_imps <- rbind(PB2v,rbind(PB1v,rbind(PAv,rbind(HAv,rbind(NPv,rbind(NAv,rbind(M1v,NS1v))))))) %>% select(-X)
write.csv(var_imps,"Varimp/VarImps.csv",row.names = F)
```

```{r}
var_imps <- read.csv("Varimp/VarImps.csv")

# ggplot(subset(var_imps,AUC_loss>0),aes(AUC_loss,colour = Protein))+ geom_density()+ ylim(0,500)
ggplot(subset(var_imps,AUC_loss>0.001),aes(AUC_loss,colour = Protein))+ geom_density()

ggplot(subset(var_imps,AUC_loss>0),aes(AUC_loss,colour = Protein))+ geom_density()+facet_wrap(~str_split_i(var,"_",1))+ ylim(0,500)

ggplot(subset(var_imps,AUC_loss>0),aes(AUC_loss,colour = Protein))+ geom_density()+facet_wrap(~str_split_i(var,"_",1),scales="free_y")
```

```{r}
ggplot(subset(var_imps,var %>% startsWith("PAAC_Xc2.lambda2")),aes(Protein,AUC_loss,colour = Protein,size=Protein=="01PB2"|Protein=="03PA"))+ geom_point()+geom_hline(yintercept = 0,colour="grey20",linetype="longdash")+ theme(legend.position = "none")+labs(title="PAAC_Xc2.lambda2")
ggsave("Figures & Presentables/VarImp1.pdf",width=33,height=19,units = "cm")

ggplot(subset(var_imps,var %>% startsWith("PAAC_Xc1.Q")),aes(Protein,AUC_loss,colour = Protein,size=Protein=="05NP"|Protein=="03PA"))+ geom_point()+geom_hline(yintercept = 0,colour="grey20",linetype="longdash")+ theme(legend.position = "none")+labs(title="PAAC_Xc1.Q")
ggsave("Figures & Presentables/VarImp2.pdf",width=33,height=19,units = "cm")

ggplot(subset(var_imps,var %>% startsWith("PAAC_Xc1.E")),aes(Protein,AUC_loss,colour = Protein,size=Protein=="05NP"|Protein=="06NA"))+ geom_point()+geom_hline(yintercept = 0,colour="grey20",linetype="longdash")+ theme(legend.position = "none")+labs(title="PAAC_Xc1.E")
ggsave("Figures & Presentables/VarImp3.pdf",width=33,height=19,units = "cm")

ggplot(subset(var_imps,var %>% startsWith("CTDD_hydrophobicity_ENGD860101.1.residue25")), aes(Protein,AUC_loss,colour = Protein,size=Protein=="05NP"|Protein=="07M1"))+ geom_point()+geom_hline(yintercept = 0,colour="grey20",linetype="longdash")+ theme(legend.position = "none")+labs(title="CTDD_hydrophobicity_ENGD860101.1.residue25")
ggsave("Figures & Presentables/VarImp4.pdf",width=33,height=19,units = "cm")

ggplot(subset(var_imps,var %>% startsWith("CTDC_normwaalsvolume.G2")),aes(Protein,AUC_loss,colour = Protein,size=Protein=="06NA"|Protein=="07M1"))+ geom_point()+geom_hline(yintercept = 0,colour="grey20",linetype="longdash")+ theme(legend.position = "none")+labs(title="CTDC_normwaalsvolume.G2")
ggsave("Figures & Presentables/VarImp5.pdf",width=33,height=19,units = "cm")
```


# Phylogenetics

## BayesTraits
```{r}

```


## BEAST
```{r}
subset(reference_table,(UID %in% rep_uids$UID) & (Protein %in% PRT)) %>% group_by(UID,Protein) %>% summarise(ntAcc) %>% pivot_wider(names_from = Protein, values_from = ntAcc) %>% write.csv("BEAST/Timesheet.csv",row.names = F)

# subset(reference_table, Protein %in% PRT) |>
#   dplyr::summarise(n = dplyr::n(), .by = c(UID, Protein)) |>
#   dplyr::filter(n > 1L)`
```

```{r}
read.csv("/Users/jordanbone/Documents/GitHub/IAV-Mach-Learning/BEAST/Acc_meta/ncbi_dataset/data/data_report.jsonl",sep="{",header = F)
```

